### PR TITLE
Jesse removal miss block

### PR DESCRIPTION
**Pull request by JesseAbram**

### CODE REVIEW

File: pallets/collator-selection/src/tests.rs [hunk 1]
```diff
@@ -247,7 +260,48 @@ fn session_management_works() {
...
+		// no one is booted
+		assert_eq!(CollatorSelection::candidates().len(), 1);
+
+		initialize_to_block(6);
+
+		// 2 new candidates one with authored block
+		assert_eq!(CollatorSelection::candidates().len(), 1);
+
+		println!("${:?}", CollatorSelection::candidates());
+		println!("${:?}", CollatorSelection::invulnerables());
```
Comment (kianenigma) [inline pallets/collator-selection/src/tests.rs:285]: you are mixing JS and Rust 😬

```suggestion
		println!("{:?}", CollatorSelection::candidates());
		println!("{:?}", CollatorSelection::invulnerables());
```

Comment (JesseAbram) [reply pallets/collator-selection/src/tests.rs:285]: hahah, just wait until you start to see solidity creeping in ;)

File: pallets/collator-selection/src/benchmarking.rs [hunk 2]
```diff
@@ -144,6 +150,22 @@ benchmarks! {
 	} verify {
 		assert!(T::Currency::free_balance(&author) > 0u32.into());
 	}
+
+	// worse case is on new session.
+	new_session {
+		let c in 1 .. T::MaxCandidates::get();
+		<CandidacyBond<T>>::put(T::Currency::minimum_balance());
+		<DesiredCandidates<T>>::put(c);
+		register_candidates::<T>(c);
```
Comment (JesseAbram) [inline pallets/collator-selection/src/benchmarking.rs:159]: The worst case scenerio would be if everyone got booted, need to fix

File: pallets/collator-selection/src/lib.rs [hunk 3]
```diff
@@ -331,14 +331,35 @@ pub mod pallet {
 			T::PotId::get().into_account()
 		}
 
+		pub fn try_remove_candidate(who: &T::AccountId) -> Result<usize, DispatchError> {
```
Comment (kianenigma) [inline pallets/collator-selection/src/lib.rs:334]: pub items should ideally be documented, specially for this case where it is mutating state.

File: pallets/collator-selection/src/lib.rs [hunk 4]
```diff
@@ -331,14 +331,35 @@ pub mod pallet {
...
 		pub fn assemble_collators() -> Vec<T::AccountId> {
 			let mut collators = Self::invulnerables();
+			let boot_block = Self::boot_block();
 			collators.extend(
-				Self::candidates().into_iter().map(|c| c.who).collect::<Vec<_>>(),
+				Self::candidates().into_iter().filter_map(|c| {
+					if c.last_block > boot_block {
+						Some(c.who)
+					} else {
+						let _ = Self::try_remove_candidate(&c.who);
```
Comment (kianenigma) [inline pallets/collator-selection/src/lib.rs:356]:
```suggestion
						let outcome = Self::try_remove_candidate(&c.who);
						if let Err(why) = outcome { 
							log::warn!("..."); 
							debug_assert!(false, "expanation...");
						}
```

Comment (kianenigma) [reply pallets/collator-selection/src/lib.rs:356]: or if you feel confident enough that `try_remove_candidate` can never fail, you can `.expect`, but I don't recommend this.

File: pallets/collator-selection/src/lib.rs [hunk 5]
```diff
@@ -355,9 +376,15 @@ pub mod pallet {
 			// `reward` is half of treasury account, this should never fail.
 			let _success = T::Currency::transfer(&treasury, &author, reward, KeepAlive);
 			debug_assert!(_success.is_ok());
-
+			<Candidates<T>>::mutate(|candidates| {
+				if candidates.len() > 0 {
+				let index: usize = candidates.iter().position(|candidate| candidate.who == author).unwrap_or_default();
```
Comment (kianenigma) [inline pallets/collator-selection/src/lib.rs:381]: faffy indentation

Comment (kianenigma) [inline pallets/collator-selection/src/lib.rs:381]:
```suggestion
				let index: usize = candidates.iter().position(|candidate| candidate.who == author).unwrap_or_default();
```

If the candidate is not found, why should we remove the first one? We should skip if it does not exists.

Note that the author might actually no longer be a candidate in our code, because of session buffering. So it is totally legit that an author is no longer candidate in our pallet and should be handled better.

File: pallets/collator-selection/src/lib.rs [hunk 6]
```diff
@@ -355,9 +376,15 @@ pub mod pallet {
...
+					candidates[index].last_block = frame_system::Pallet::<T>::block_number();
+				}
```
Comment (kianenigma) [inline pallets/collator-selection/src/lib.rs:383]:
```suggestion
				if let Some(found) = candidates.iter_mut().find(|candidate| candidate.who == author) {
					found.last_block = frame_system::Pallet::<T>::block_number();
				}
```

File: pallets/collator-selection/src/lib.rs [hunk 7]
```diff
@@ -355,9 +376,15 @@ pub mod pallet {
...
+			});
+			let candidates_len = Self::candidates().len();
```
Comment (kianenigma) [inline pallets/collator-selection/src/lib.rs:385]: you can return the length from `<Candidates<T>>::mutate(||...)`

File: pallets/collator-selection/src/lib.rs [hunk 8]
```diff
@@ -375,7 +402,14 @@ pub mod pallet {
 				index,
 				<frame_system::Pallet<T>>::block_number(),
 			);
-			Some(Self::assemble_collators())
+			let candidates_len = Self::candidates().len();
+			let result = Some(Self::assemble_collators());
```
Comment (kianenigma) [inline pallets/collator-selection/src/lib.rs:406]: you should not wrap it in `Some` here, then you need to unwrap in the line below, this is like making things harder for yourself for no reason 😄

File: pallets/collator-selection/src/lib.rs [hunk 9]
```diff
@@ -375,7 +402,14 @@ pub mod pallet {
...
+			let removed = candidates_len - result.clone().unwrap_or_default().len();
+			frame_system::Pallet::<T>::register_extra_weight_unchecked(
+				T::WeightInfo::new_session(candidates_len as u32, removed as u32),
```
Comment (kianenigma) [inline pallets/collator-selection/src/lib.rs:409]: Can you explain how the weight of the new session is now being calculated?

File: runtime/statemine/src/lib.rs [hunk 10]
```diff
@@ -736,15 +736,15 @@ pub type Executive = frame_executive::Executive<
...
-	// impl sp_consensus_aura::AuraApi<Block, AuraId> for Runtime {
-	// 	fn slot_duration() -> sp_consensus_aura::SlotDuration {
-	// 		sp_consensus_aura::SlotDuration::from_millis(Aura::slot_duration())
-	// 	}
-
-	// 	fn authorities() -> Vec<AuraId> {
-	// 		CollatorSelection::Collators()
-	// 	}
-	// }
+	impl sp_consensus_aura::AuraApi<Block, AuraId> for Runtime {
```
Comment (kianenigma) [inline runtime/statemine/src/lib.rs:739]: cc @joepetrowski @bkchr all of what we did for aura support was to just add these few lines here.

File: pallets/collator-selection/src/tests.rs [hunk 11]
```diff
@@ -227,18 +240,17 @@ fn session_management_works() {
 		assert_eq!(SessionChangeBlock::get(), 0);
 		assert_eq!(SessionHandlerCollators::get(), vec![1, 2]);
 
-		// add a new collator
-		assert_ok!(CollatorSelection::register_as_candidate(Origin::signed(3)));
```
Comment (kianenigma) [inline pallets/collator-selection/src/tests.rs:231]: why did this have to change?

File: pallets/collator-selection/src/tests.rs [hunk 12]
```diff
@@ -247,7 +259,31 @@ fn session_management_works() {
...
+fn boot_mechanism() {
+	new_test_ext().execute_with(|| {
+		// add a new collator
+		assert_ok!(CollatorSelection::register_as_candidate(Origin::signed(3)));
+		assert_ok!(CollatorSelection::register_as_candidate(Origin::signed(4)));
+		assert_eq!(CollatorSelection::candidates().len(), 2);
+		initialize_to_block(21);
+		assert_eq!(CollatorSelection::boot_block(), 20);
+		assert_eq!(SessionChangeBlock::get(), 20);
+		// 4 authored this block, gets to stay 3 was booted
```
Comment (kianenigma) [inline pallets/collator-selection/src/tests.rs:276]: is "being booted" == "being kicked", terminology-wise?

**Review by kianenigma — COMMENTED** [review]
Have some comments, but mostly looks good.

File: pallets/collator-selection/src/lib.rs (hunk 9 above)
Comment (JesseAbram) [reply pallets/collator-selection/src/lib.rs:412]: yup, I take the worse case scenerio which is the max size collators with all of them being removed. I test for that in the benchmark (c... and r... T::MaxCandidates). Then at the end of the call I pass through how big the candidates vec was and how many were removed

File: pallets/collator-selection/src/tests.rs (hunk 12 above)
Comment (JesseAbram) [reply pallets/collator-selection/src/tests.rs:276]: ya, tbh I was gonna bring up that I was having trouble naming this variable lol

File: runtime/statemine/src/lib.rs (hunk 10 above)
Comment (bkchr) [reply runtime/statemine/src/lib.rs:739]: This is not complete. 

Next week when I'm back, I will be merging the aura pr. Then we can do it here properly.

File: pallets/collator-selection/src/tests.rs (hunk 11 above)
Comment (JesseAbram) [reply pallets/collator-selection/src/tests.rs:231]: ahhh so because of the booting It changed your tests significantly because now the adding the extra collator in was being booted before the bottom half of the tests, so I was like hmm if I just add it higher with another candidate this would keep the same spirit of your tests and deal with the removal

File: pallets/collator-selection/src/tests.rs (hunk 12 above)
Comment (kianenigma) [reply pallets/collator-selection/src/tests.rs:276]: I would call kicking or yanking.

File: pallets/collator-selection/src/benchmarking.rs [hunk 13]
```diff
@@ -50,12 +51,23 @@ fn assert_last_event<T: Config>(generic_event: <T as Config>::Event) {
...
-fn register_candidates<T: Config>(count: u32) {
+fn register_candidates<T: Config>(count: u32, removals: u32) {
 	let candidates = (0..count).map(|c| account("candidate", c, SEED)).collect::<Vec<_>>();
 	assert!(<CandidacyBond<T>>::get() > 0u32.into(), "Bond cannot be zero!");
+	let mut i = 0;
 	for who in candidates {
 		T::Currency::make_free_balance_be(&who, <CandidacyBond<T>>::get() * 2u32.into());
-		<CollatorSelection<T>>::register_as_candidate(RawOrigin::Signed(who).into()).unwrap();
+
+		let last_block = if count - i > removals {10u32.into()} else {0u32.into()};
```
Comment (kianenigma) [inline pallets/collator-selection/src/benchmarking.rs:61]:
```suggestion
		let last_block = if count - i > removals { 10u32.into() } else { 0u32.into() };
```

File: pallets/collator-selection/src/lib.rs [hunk 14]
```diff
@@ -330,15 +330,40 @@ pub mod pallet {
...
 		/// Assemble the current set of candidates and invulnerables into the next collator set.
 		///
 		/// This is done on the fly, as frequent as we are told to do so, as the session manager.
 		pub fn assemble_collators() -> Vec<T::AccountId> {
 			let mut collators = Self::invulnerables();
+			let boot_block = Self::boot_block();
 			collators.extend(
-				Self::candidates().into_iter().map(|c| c.who).collect::<Vec<_>>(),
+				Self::candidates().into_iter().filter_map(|c| {
+					if c.last_block > boot_block {
```
Comment (kianenigma) [inline pallets/collator-selection/src/lib.rs:353]: I probably recommended this myself, but this is wrong. What you need is:

```
let now = system::now();
let boot_threshold = T::boot();
let last = c.last_block;
let since_last = now - last;

if since_last > boot { ... }
```

File: pallets/collator-selection/src/tests.rs (hunk 11 above)
Comment (kianenigma) [reply pallets/collator-selection/src/tests.rs:231]: What about make booting optional, through either of:

- its type would be `Get<Option<_>>` and if it is `None`, we never boot.
- At least in the context of tests, if you want to never boot, we can just set its value to `Bounded::max`

File: pallets/collator-selection/src/benchmarking.rs (hunk 13 above)
Comment (kianenigma) [reply pallets/collator-selection/src/benchmarking.rs:61]: Also, the relationship between this 10 and the `removal` can probably be enhanced. Looking at your benchmark, you want to have some candidates kicked at each new session. I think a nicer way of doing this is:

- leave this function as-is, make it always set `last_block` to `now`.
- in the benchmark that is interested in this, manually set `last_block` of `r` of these folks to `now - (boot + 1)` (as opposed to the magic 0 and 10).

**Review by kianenigma — COMMENTED** [review]
On a good track, but I have some further comments 🙈

File: pallets/collator-selection/src/lib.rs (hunk 14 above)
Comment (JesseAbram) [reply pallets/collator-selection/src/lib.rs:353]: So no I don't think this is, What is happening is lower in the function KickBlock (oh ya boot block, I just renamed it will push soon) gets set (it would start as 0). Then a session happens block is now X, if a collator produced a block in the time they will be greater then kickBlock else they will be kicked then kick block gets moved up again

File: pallets/collator-selection/src/tests.rs (hunk 11 above)
Comment (JesseAbram) [reply pallets/collator-selection/src/tests.rs:231]: hmmm idk if I want to change the logic of the pallet for the tests, I think a less intrusive way is that I create a function that changes the last_block manually to a much higher block so it never boots, then I can put your test back to the original

Comment (JesseAbram) [reply pallets/collator-selection/src/tests.rs]: ok after reverting it back turns out after some changes 3 now authors the block first and does not get booted, test reverted back to original, no change needed

File: pallets/collator-selection/src/lib.rs [hunk 15]
```diff
@@ -375,7 +405,16 @@ pub mod pallet {
 				index,
 				<frame_system::Pallet<T>>::block_number(),
 			);
-			Some(Self::assemble_collators())
+
+			let candidates_len_before = Self::candidates().len();
+			let result = Self::assemble_collators();
```
Comment (kianenigma) [inline pallets/collator-selection/src/lib.rs:410]: This is not a required change but I want to point out that assemble_collators is having side effects which makes the code here a bit hard to follow. 

reading the candidates once, and getting passing it through two functions where we:
1. kick stale ones
2. prefix with invulnerables

```
let mut candidates = Self::candidates();
Self::kick_stale_candidates(&mut candidates);
let collators = Self:: assemble_collators(candidates);
```

is somewhat cleaner. WDYT?

File: pallets/collator-selection/src/lib.rs (hunk 14 above)
Comment (kianenigma) [reply pallets/collator-selection/src/lib.rs]: I see, then it is semantically the same as why I recommended, right? except, in your way we have to read and write into a storage item, but you can do the same if you make `KickBlock` a config `Get`. Or am I missing something?

File: pallets/collator-selection/src/lib.rs (hunk 15 above)
Comment (JesseAbram) [reply pallets/collator-selection/src/lib.rs:410]: ya I think so, actually if I return just a vec of accounts from kick stale then I can get rid of the map in assemble collators too

File: pallets/collator-selection/src/lib.rs (hunk 14 above)
Comment (JesseAbram) [reply pallets/collator-selection/src/lib.rs]: ya, but to avoid config confustion Id want to make the kick_block the session block, this is becuase this only runs on a new session, which makes sense to me because that is when we set new authoritites.

File: pallets/collator-selection/src/lib.rs [hunk 16]
```diff
@@ -330,17 +330,46 @@ pub mod pallet {
...
-		pub fn assemble_collators() -> Vec<T::AccountId> {
+		pub fn assemble_collators(candidates: Vec<T::AccountId>) -> Vec<T::AccountId> {
 			let mut collators = Self::invulnerables();
 			collators.extend(
-				Self::candidates().into_iter().map(|c| c.who).collect::<Vec<_>>(),
+				candidates.into_iter().collect::<Vec<_>>(),
 			);
 			collators
 		}
+		/// Kicks out and candidates that did not produce a block in the last session.
```
Comment (kianenigma) [inline pallets/collator-selection/src/lib.rs:355]:
```suggestion
		/// Kicks out and candidates that did not produce a block since the last call to this function. Given that this is called once per session, this essentially ensures that the each collator must author one block per session. 
```


and if we have more collators per session length? :D

**Review by kianenigma — APPROVED** [review]
I would have done some details a bit differently, but looks correct 👍

File: pallets/collator-selection/src/lib.rs (hunk 16 above)
Comment (JesseAbram) [reply pallets/collator-selection/src/lib.rs:353]: With the last commit this now can be a multiple of sessions and can be extended :)

File: pallets/collator-selection/src/lib.rs [hunk 17]
```diff
@@ -375,7 +408,18 @@ pub mod pallet {
 				index,
 				<frame_system::Pallet<T>>::block_number(),
 			);
-			Some(Self::assemble_collators())
+
+			let candidates = Self::candidates();
+			let candidates_len_before = candidates.len();
+			let active_candidates = Self::kick_stale_candidates(candidates);
+			let result = Self::assemble_collators(active_candidates.clone());
+			let removed = candidates_len_before - active_candidates.len();
```
Comment (kianenigma) [inline pallets/collator-selection/src/lib.rs:416]:
```suggestion
			let active_candidates = Self::kick_stale_candidates(candidates);
			let active_candidates_len = active_candidates.len();
			let result = Self::assemble_collators(active_candidates);
			let removed = candidates_len_before - active_candidates_len;
```

cloning the length is much cheaper than cloning the whole thing.
